tfjs-backend-webgpu/src/scatter_optimized_webgpu.ts

```diff
@@ -64,45 +66,50 @@ export class ScatterOptimizedProgram implements WebGPUProgram {
     const strideString = this.sliceDimGreaterThanOne ? 'uniforms.strides[j]' :
                                                        'uniforms.strides';
 
-    let updatesString = '';
     let outCoordsString = '';
     let getUpdatesCoordsFromFlatIndex = '';
-    if (this.updatesRank === 1) {
-      updatesString = 'coords[0]';
+    if (this.dispatchLayout.x.length === 1) {
       outCoordsString = 'flattenedIndex';
       getUpdatesCoordsFromFlatIndex = `
       fn getUpdatesCoordsFromFlatIndex(index : i32) -> i32 {
         return index;
       }
       `;
-    } else if (this.updatesRank === 2) {
-      updatesString = 'coords[0], coords[1]';
+    } else if (this.dispatchLayout.x.length === 2) {
       outCoordsString = 'vec2<i32>(flattenedIndex, coords[1])';
       getUpdatesCoordsFromFlatIndex = `
       fn getUpdatesCoordsFromFlatIndex(index : i32) -> vec2<i32> {
-        let d0 = index / uniforms.updatesShape[1];
-        let d1 = index - d0 * uniforms.updatesShape[1];
+        let d0 = index / uniforms.outShape[1];
+        let d1 = index - d0 * uniforms.outShape[1];
         return vec2<i32>(d0, d1);
       }
       `;
     }
-    const updatesSnippet = `getUpdates(${updatesString})`;
+    const updatesString =
+        Array.from({length: this.updatesRank}, (_, idx) => `coords[${idx}]`);
+    const updatesSnippet = `getUpdates(${updatesString.join(', ')})`;
 
-    // atomicAdd only supports uint/int type. For float, we use
-    // atomicCompareExchangeWeak to simulate.
-    const atomicAddSnippet = this.type === 'int32' ?
-        `atomicAdd(&(result[flatIndex]), i32(updateValue));` :
-        `
-     var oldValue = atomicLoad(&(result[flatIndex]));
-     var exchanged = false;
-     for (; !exchanged;) {
-       let newValueF32 = bitcast<f32>(oldValue) + updateValue;
-       let newValue = bitcast<i32>(newValueF32);
-       let res = atomicCompareExchangeWeak(&(result[flatIndex]), oldValue, newValue);
-       oldValue = res.old_value;
-       exchanged = res.exchanged;
-     }
-     `;
+    const atomicRMW = (ptr: string, val: string) => {
+      const atomicAddSnippet = `
+        {
+          var oldBits = 0;
+          var newBits = bitcast<i32>(${val});
+          loop {
+            let info = atomicCompareExchangeWeak(${ptr}, oldBits, newBits);
+            if (info.exchanged) {
+              break;
+            }
+            oldBits = info.old_value;
+            let oldValue =
+                bitcast<${mapToWgslTypes(this.type, false)}>(oldBits);
+            let newValue = oldValue + (${val});
+            newBits = bitcast<i32>(newValue);
+          }
+        }
+      `;
+      const atomicStoreSnippet = `atomicStore(${ptr}, bitcast<i32>(${val}));`;
```

Comment:
Why not just store `0` so that we don't need to upload a zero tensor as input?

Reply:
This fits my taste better: the kernel does something (accumulation or overwriting) according to indices and values given by the user.